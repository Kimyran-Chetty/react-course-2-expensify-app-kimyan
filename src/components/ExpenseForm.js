import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';
//import 'react-dates/lib/css/_datepicker.css'

//Rudemetry way to do dates is by using the date class which is tedious and complex eg:
//const date =new Date();

//We will always use moment :
// const now  = moment();
// console.log(now.format('MMM, Do, YYYY'));


export default class ExpenseForm extends React.Component{

    constructor(props){   //We now use a constructor to get access to the props which contains the state (of expenses)
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount/100).toString() : '',
            createdAt : props.expense ? moment(props.expense.createdAt) : moment(),
            calenderFocused : false,
            error: ''
    
        }
    }

    onDescriptionChange = (e) => {

        const description = e.target.value;
        this.setState(() => ({description}))

    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}))
    }

    onAmountChange = (e) => {

        const amount = e.target.value;
        if( !amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({amount}))
        }
    }

    onDateChange = (createdAt) => {

        if(createdAt){
            this.setState(()=> ({createdAt}))
        }
    }

    onFocusChange = ({focused}) => {
        this.setState(() => ({calenderFocused: focused}))
    }

    //Changed for test
    // onFocusChange = (focused) => {
    //     this.setState(() => ({calanderFocused: focused}))
    // }

    onSubmit = (e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.amount){
            //Set error state equal to 'Please provide description .and amount.'
            this.setState(() =>({error: 'Please provide description and amount.'}))
        }
        else{
            //Clear error 
            this.setState(() =>({error: ''}))
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) *100 ,  //Converting text amount to actual decimal
                createdAt: this.state.createdAt.valueOf(), // Method of moment to convent to millisecond timestamp
                note: this.state.note
            });

            console.log('submitted')
        }
    }
    
    render() {
        return (
            <div>

            {this.state.error && <p>{this.state.error}</p>}

                <form onSubmit = {this.onSubmit}>
                    <input
                    type="text"
                    placeholder = "Description"
                    autoFocus                     //Doesnt work
                    value = {this.state.description}
                    onChange = {this.onDescriptionChange}
                    />

                    <input
                    type="text"
                    placeholder = "Amount"
                    value = {this.state.amount}
                    onChange = {this.onAmountChange}
                    />

                    <SingleDatePicker
                    date = {this.state.createdAt}
                    onDateChange = {this.onDateChange}
                    focused = {this.state.calenderFocused}
                    onFocusChange = {this.onFocusChange}
                    numberOfMonths = {1}
                    isOutsideRange = {() => false}   // All days from past available // {(day) => {logic}} --- something like this could be used to valaidate if days are selectable

                    
                    />

                    <textarea
                    placeholder = "Add a note for your expense (optional)" 
                    value = {this.state.note}
                    onChange = {this.onNoteChange}
                    >
                    </textarea>

                    <button>Add Expense</button>



                </form>
            </div>
        )
    }
}