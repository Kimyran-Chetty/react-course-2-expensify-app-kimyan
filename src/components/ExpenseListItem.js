import React from 'react';
import {Link} from 'react-router-dom';

//My solution
const ExpenseListItem = (props) => {  
    return (
        <div>
            <Link to = {`/edit/${props.item.id}`}>
            <h3>{props.item.description}</h3>
            </Link>
            
            <p> Amount: {props.item.amount} CreatedAt: {props.item.createdAt}</p>
            
        </div>
    )
}

// solution using spread and deconstruction 
// const ExpenseListItem = ({dispatch,id,description, amount, createdAt}) => {  // Was ExpenseListItem = (props) , however we deconstructed the props object
//     return (
//         <div>
//             <h3>{description}</h3>
//             <p> Amount: {amount} CreatedAt: {createdAt}</p>
//             <button onClick = {() => {
//                  props.dispatch(removeExpense(id))
//             }}>
//             Remove </button>
//         </div>
//     )
// }

//export default connect()(ExpenseListItem);

export default ExpenseListItem;