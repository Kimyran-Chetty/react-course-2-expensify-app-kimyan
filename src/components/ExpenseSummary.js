import React from 'react';
import {connect} from 'react-redux';
import getVisbleExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

// export const ExpenseSummary = (props) => {
//     return (
//             <div>
//             { props.expensesVisible.length > 0 &&
//                 <div>
//                 <h2>
//                   Viewing {props.expensesVisible.length} {props.expensesVisible.length === 1? 'expense': 'expenses'}
//                 </h2>
//                 <h2>
//                   Totalling: {numeral(props.expensesTotal/100).format('$0,0.00')}
//                 </h2>
//                 </div>
//             }
//             </div>

//     )
// }

export const ExpenseSummary = ({expensesCount,expensesTotal}) => {

  const expenseWord = expensesCount === 1? 'expense': 'expenses';
  const formattedExpensesTotal = numeral(expensesTotal/100).format('$0,0.00');

  return (
          <div>
          { expensesCount > 0 &&
              <div>
              <h1>
                Viewing {expensesCount} {expenseWord}, totalling: {formattedExpensesTotal}
              </h1>
             
              </div>
          }
          </div>

  )
}



const mapStateToProps = (state) => {

  const expensesVisible = getVisbleExpenses(state.expenses,state.filters);

    return{
        
        expensesCount: expensesVisible.length,
        expensesTotal: getExpensesTotal(expensesVisible)
    }
    
}

export default connect (mapStateToProps)(ExpenseSummary);

