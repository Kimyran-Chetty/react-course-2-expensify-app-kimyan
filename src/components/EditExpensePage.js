import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import {editExpense,removeExpense} from '../actions/expenses';

export class EditExpensePage extends React.Component{

    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
        console.log('updated',expense)
    }


    onClick = () => {
        this.props.removeExpense(this.props.expense.id);
        this.props.history.push('/');
    }


    render() {
        return (
                    <div>
                    
                    <ExpenseForm
                    expense = {this.props.expense}
                    onSubmit= {this.onSubmit}
                    />

                    <button onClick = {this.onClick}>
                    Remove </button>

                </div>
        )
    }
}

// const EditExpensePage = (props) => {
//     //console.log(props);
//     return (
//         <div>
            
//             <ExpenseForm
//             expense = {props.expense}
//             onSubmit= {(expense) => {
//                 props.dispatch(editExpense(props.expense.id, expense));
//                 props.history.push('/');
//                 console.log('updated',expense)
//             }}
//             />

//             <button onClick = {() => {
//                 props.dispatch(removeExpense(props.expense.id));
//                 props.history.push('/');
//             }}>
//             Remove </button>

//         </div>
//     );

// };

const mapStateToProps = (state, props) => {
    return{
        expense: state.expenses.find((expense) => {
            if(expense.id === props.match.params.id){
                return true;
            }
            else{
                return false;
            }
        })
    };
}

const mapDispatchToProps = (dispatch) => {
    return {
        editExpense: (id, expense) => dispatch(editExpense(id, expense)),
        removeExpense: (id) => dispatch(removeExpense(id))
        }
}


export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage);
//export default EditExpensePage;