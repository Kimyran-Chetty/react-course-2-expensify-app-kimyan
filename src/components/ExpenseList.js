import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisbleExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (

    <div>
        {props.expenses.length === 0 ? (

            <p>No Expenses</p>

            ) : 
            
            (
                props.expenses.map((expense) => {
                    
                //My solution
                return <ExpenseListItem 
                        key = {expense.id}
                        item ={expense}
                        />

                }

                //New Solution using spread
                // {
                // return <ExpenseListItem key= {expense.id} {...expense}
                // />
                // }

                )       

            )

        }

      
       
    </div>

);

// const ConnectedExpenseList = connect((state)=>{
//     return{
//         expenses: state.expenses
//     };
// })(ExpenseList);

// export default ConnectedExpenseList;

//Common practise for the above code:

const mapStateToProps = (state)=>{
    return{
        //expenses: state.expenses,
        expenses: getVisbleExpenses(state.expenses,state.filters)
     
    };
}

export default connect(mapStateToProps)(ExpenseList);