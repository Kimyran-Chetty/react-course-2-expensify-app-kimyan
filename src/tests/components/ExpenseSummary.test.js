import React from 'react'
import {shallow} from 'enzyme';
import {ExpenseSummary} from '../../components/ExpenseSummary';


test('should render ExpenseForm', () => {
    const wrapper = shallow(<ExpenseSummary expensesCount={1}  expensesTotal={235}/>);
    expect(wrapper).toMatchSnapshot();
})

test('should render ExpenseForm', () => {
    const wrapper = shallow(<ExpenseSummary expensesVisible={23}  expensesTotal={245678345}/>);
    expect(wrapper).toMatchSnapshot();
})