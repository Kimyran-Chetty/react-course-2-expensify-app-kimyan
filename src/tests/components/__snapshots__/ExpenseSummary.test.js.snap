// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseForm 1`] = `
<div>
  <div>
    <h1>
      Viewing 
      1
       
      expense
      , totalling: 
      $2.35
    </h1>
  </div>
</div>
`;

exports[`should render ExpenseForm 2`] = `<div />`;
